polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs

```diff
@@ -0,0 +1,1047 @@
+	/// The candidate was already known.
+	CandidateAlreadyKnown(CandidateHash),
+}
+
+/// Stores candidates and information about them such as their relay-parents and their backing
+/// states.
+#[derive(Clone)]
+pub(crate) struct CandidateStorage {
+	// Index from head data hash to candidate hashes with that head data as a parent. Purely for
+	// efficiency when responding to `ProspectiveValidationDataRequest`s or when trying to find a
+	// new candidate to push to a chain.
+	// Even though having multiple candidates with same parent would be invalid for a parachain, it
+	// could happen across different relay chain forks, hence the HashSet.
+	by_parent_head: HashMap<Hash, HashSet<CandidateHash>>,
+
+	// Index from head data hash to candidate hashes outputting that head data. Purely for
+	// efficiency when responding to `ProspectiveValidationDataRequest`s.
+	// Even though having multiple candidates with same output would be invalid for a parachain,
+	// it could happen across different relay chain forks.
+	by_output_head: HashMap<Hash, HashSet<CandidateHash>>,
+
+	// Index from candidate hash to fragment node.
+	by_candidate_hash: HashMap<CandidateHash, CandidateEntry>,
+}
+
+impl CandidateStorage {
+	/// Create a new `CandidateStorage`.
+	pub fn new() -> Self {
+		CandidateStorage {
+			by_parent_head: HashMap::new(),
+			by_output_head: HashMap::new(),
+			by_candidate_hash: HashMap::new(),
+		}
+	}
+
+	/// Introduce a new candidate.
+	pub fn add_candidate(
+		&mut self,
+		candidate: CommittedCandidateReceipt,
+		persisted_validation_data: PersistedValidationData,
+		state: CandidateState,
+	) -> Result<CandidateHash, CandidateStorageInsertionError> {
+		let candidate_hash = candidate.hash();
+		if self.by_candidate_hash.contains_key(&candidate_hash) {
+			return Err(CandidateStorageInsertionError::CandidateAlreadyKnown(candidate_hash))
+		}
+
+		if persisted_validation_data.hash() != candidate.descriptor.persisted_validation_data_hash {
+			return Err(CandidateStorageInsertionError::PersistedValidationDataMismatch)
+		}
+
+		let entry = CandidateEntry {
+			candidate_hash,
+			parent_head_data_hash: persisted_validation_data.parent_head.hash(),
+			output_head_data_hash: candidate.commitments.head_data.hash(),
+			relay_parent: candidate.descriptor.relay_parent,
+			state,
+			candidate: ProspectiveCandidate {
+				commitments: Cow::Owned(candidate.commitments),
+				collator: candidate.descriptor.collator,
+				collator_signature: candidate.descriptor.signature,
+				persisted_validation_data,
+				pov_hash: candidate.descriptor.pov_hash,
+				validation_code_hash: candidate.descriptor.validation_code_hash,
+			},
+		};
+
+		self.by_parent_head
+			.entry(entry.parent_head_data_hash())
+			.or_default()
+			.insert(candidate_hash);
+		self.by_output_head
+			.entry(entry.output_head_data_hash())
+			.or_default()
+			.insert(candidate_hash);
+		// sanity-checked already.
+		self.by_candidate_hash.insert(candidate_hash, entry);
```

Comment:
I'd extract this into a fn called `add_entry`.

Reply:
this one is already called `add_candidate`. I don't think it helps